### PR TITLE
Feat/cw storage spec

### PR DESCRIPTION
**Pull request by ccamel**
## 📝 Purpose

This PR introduces the specification for a new smart contract, `cw-storage`, enabling the storage of arbitrary `objects` on-chain in an immutable way.

Note: The smart contract is primarily intended to store logic programs in a secure and immutable form, with the ability for programs to reference other programs to improve reusability and composability. However this smart contract is intended to be used in any other context and to be used by any other Cosmos chain.

### Basic

- *Object storage*: enables storing objects on the chain in a immutable way.
- *Object pinning*: allows users to pin objects ensuring that they are always available and cannot be removed from the chain.
- *Object unpinning*: enables users to unpin objects allowing them to be removed when necessary.
- *Object removal (forgetting)*: allows users to remove (forget) objects from the chain completely if object is not pinned anymore.

The contract specifies the essential and fundamental queries necessary to retrieve objects from the chain, with cursor-based pagination.

### Limits

The contract instance which represents a Bucket storage can include limits that define the maximum acceptable values for the bucket, such as the maximum number of objects, maximum size of an object, and maximum size of the bucket. While these values are optional, it is highly recommended to set them for security reasons, particularly in the context of a development or testing network.

## 🔮 Future

The following is under consideration for the future.

-  additional queries to support more complex use cases.
- support for metadata (bucket, objects)

### CODE REVIEW

**Comment by ccamel** [comment]
I improved the specification a bit after discussion with @bdeneux.

File: contracts/cw-storage/src/msg.rs [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    pub max_object_size: Option<Uint128>,
+    /// The maximum number of pins in the bucket for an object.
+    pub max_object_pins: Option<Uint128>,
+}
+
+/// Instantiate messages
+#[cw_serde]
+pub struct InstantiateMsg {
+    /// The name of the bucket.
+    pub bucket: String,
```
Comment (amimart) [inline contracts/cw-storage/src/msg.rs:30]: What's the purpose of the bucket name field in the instantiate message?

Comment (ccamel) [reply contracts/cw-storage/src/msg.rs:30]: Well, the name is just to give a name to the smart contract instancerepresenting a bucket, and which could be queried later with some metadata...

Comment (ccamel) [reply contracts/cw-storage/src/msg.rs:30]: I'll add the `bucket` query to allow the retrieval of bucket info (name + limits), if you agree.

Comment (ccamel) [reply contracts/cw-storage/src/msg.rs:30]: It's done. I think everything should be ok now.

**Comment by ccamel** [comment]
@amimart Regarding the ObjectId hash algorithm, I think we should consider algorithms that provide sufficient protection against collisions while not imposing excessive computational costs. But ideally I would like to consider the hashing algo as a configurable option of the bucket. Why not start with [SHA-256](https://en.wikipedia.org/wiki/SHA-2) by default?

For the Cursor representation, I absolutely love the idea of using a [base-58](https://digitalbazaar.github.io/base58-spec/) encoding you suggested! It's a brilliant idea! 😋

**Comment by amimart** [comment]
> @amimart Regarding the ObjectId hash algorithm, I think we should consider algorithms that provide sufficient protection against collisions while not imposing excessive computational costs. But ideally I would like to consider the hashing algo as a configurable option of the bucket. Why not start with [SHA-256](https://en.wikipedia.org/wiki/SHA-2) by default?
> 
> For the Cursor representation, I absolutely love the idea of using a [base-58](https://digitalbazaar.github.io/base58-spec/) encoding you suggested! It's a brilliant idea! 😋

Totally agree with you about making the hash algorithm configurable, and the SHA256 seems a good compromise regarding safety and cost efficiency 😉

**Review by amimart — APPROVED** [review]
I think this is sufficiently precise to go on a first implementation, great work! 💪

**Review by bdeneux — APPROVED** [review]
That seems very good, thanks ! Let's start coding 🚀🚀

**Comment by ccamel** [comment]
Thanks guys! Let's go.

**Pull request merged**